Build Docker image and push to Dockerhub

guidopetri opened the pull request:
As mentioned in [#1051](https://github.com/stringer-rss/stringer/issues/1051#issuecomment-1679387966), the CircleCI docker build step isn't working, and we want to move it to a Github Action instead. This PR has two commits:

- first, revert the CircleCI additions so that the CircleCI tests file is clean (via reverting the two commits from #1065)
- second, add a build+push to dockerhub github action.

This will require two secrets that I *believe* can be set in the repository settings somewhere: one for a `DOCKERHUB_USERNAME` and another for a `DOCKERHUB_TOKEN`. I also went ahead and added both an amd64 and an arm64 build here, and am hoping that it just works (tm).

I also once again set the build and push to only happen when `main` is pushed to / gets new commits, but I think we can change that to work only on this branch from my fork if we want to test before merging.

guidopetri commented:
@mockdeep hopefully it works this time!

mockdeep commented on .github/workflows/build.yaml:
In terms of naming, I think I mostly see yaml files with the `.yml` extension.

guidopetri replied:
Renamed!

mockdeep commented on .github/workflows/build.yaml:
```diff
@@ -0,0 +1,57 @@
+          context: .
+          file: ./Dockerfile
+          platforms: linux/amd64,linux/arm64
+          push: true
```
I think we might be able to build on *every* Github push and only push to Docker only on `main` with something like:

```suggestion
          push: ${{ github.ref_name == 'main' }}
```

mockdeep commented on .github/workflows/build.yaml:
```diff
@@ -0,0 +1,57 @@
+name: Build Docker image and push to Dockerhub
+
+on:
+  push:
+    branches:
+      - "main"
```
I'm wondering if we can have this build on *every* branch/commit and only push to Docker on `main`. That way we can be sure this builds, and subsequent changes don't break it, before merging.

```suggestion
on:
  push
```

mockdeep commented on .github/workflows/build.yaml:
```diff
@@ -0,0 +1,57 @@
+        id: login
+        uses: docker/login-action@v2
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
```
I've set these secrets, so hopefully that'll work.

mockdeep commented on .github/workflows/build.yaml:
```diff
@@ -0,0 +1,57 @@
+          images: ${{ env.IMAGE_NAME }}
+          tags: type=raw, value=latest
+
+      - name: Log in to Docker Hub
```
If we're building this on every branch, I think we'll need to skip this step unless we're on `main`:

```suggestion
      - name: Log in to Docker Hub
        if: ${{ github.ref_name == 'main' }}
```

mockdeep reviewed (commented):
Thanks so much for this! I think this will help a lot with environment issues.

guidopetri commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,56 @@
+          tags: |
+            ${{ steps.get-tag-latest.outputs.tags }}
+            ${{ steps.get-tag-sha.outputs.tags }}
+          cache-from: type=gha
```
I *believe* that this `cache-from` will cache to/from the Github Actions environment, so there's no need to pull an image locally, but I'm not 100% confident.

guidopetri commented:
Updated the PR with your suggestions and rebased them into the second commit. Should I keep the `run-tests` name on CircleCI instead of reverting to `build`?

mockdeep commented:
Let's give it a try! I'm wondering why we don't see the image building on this PR, but maybe we need to merge it before it will run.

> Should I keep the `run-tests` name on CircleCI instead of reverting to `build`

I think `build` is fine for now. We can always revisit later.

guidopetri commented:
It worked?!!

mockdeep commented:
Oh, I didn't even think to look. It did!

mockdeep commented:
https://hub.docker.com/repository/docker/stringerrss/stringer/general
